Add readOnly in curation page

gebbing12 opened the pull request:
Fixes https://github.com/oncokb/oncokb-pipeline/issues/236

calvinlu3 commented on src/main/webapp/app/pages/curation/collapsible/CancerTypeCollapsible.tsx:
```diff
@@ -204,19 +211,21 @@ function CancerTypeCollapsible({
             <>
               <CommentIcon id={`${cancerTypesUuid}_diagnostic_comments`} path={`${cancerTypePath}/diagnostic_comments`} />
-              <RCTButton cancerTypePath={cancerTypePath} relevantCancerTypesInfoPath={`${cancerTypePath}/diagnostic`} />
+              <RCTButton cancerTypePath={cancerTypePath} relevantCancerTypesInfoPath={`${cancerTypePath}/diagnostic`} readOnly={readOnly} />
```
Passing in the readOnly props explicitly is not required because it will be injected by mobx.

calvinlu3 reviewed (changes requested):
Thanks for making the changes!

Would you also try adding a banner to the curation page that alerts tells the user that the page is currently in read only mode?
Some text along the lines of "This page is currently in read-only mode because ___ user is currently reviewing"

gebbing12 commented on src/main/webapp/app/pages/curation/button/RCTButton.tsx:
```diff
@@ -13,9 +13,16 @@ import { Button } from 'reactstrap';
 export interface IRCTButtonProps extends StoreProps {
   cancerTypePath: string;
   relevantCancerTypesInfoPath: string; // path to dx, px, or tx
+  readOnly?: boolean;
```
I added the readOnly property to RCTButton because I thought this component might be used in multiple places simultaneously. If it’s only used in CurationPage, I’ll remove this line and use StoreProp instead.

gebbing12 commented:
> Thanks for making the changes!
> 
> Would you also try adding a banner to the curation page that alerts the user that the page is currently in read only mode? Some text along the lines of "This page is currently in read-only mode because ___ user is currently reviewing"

I have added a banner to the top of the page. Please let me know if there are any improvements needed for the text or style.
![691733353841_ pic](https://github.com/user-attachments/assets/509ccc2a-e838-4da1-96e1-9428030977ea)
ow

calvinlu3 commented:
> I have added a banner to the top of the page. Please let me know if there are any improvements needed for the text or style. ![691733353841_ pic](https://private-user-images.githubusercontent.com/127727594/392596859-509ccc2a-e838-4da1-96e1-9428030977ea.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQwOTk4MTUsIm5iZiI6MTczNDA5OTUxNSwicGF0aCI6Ii8xMjc3Mjc1OTQvMzkyNTk2ODU5LTUwOWNjYzJhLWU4MzgtNGRhMS05NmUxLTk0MjgwMzA5NzdlYS5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIxM1QxNDE4MzVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05MjBhZDliZTYwZTY1ZTNkN2ZiZDI4NzVlZDUyZTg2YjY2YjQ2ZTJjYmU0MzFiNjE5MmQ3ZTg4ZTgzMjJiY2EzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.bwI98oCVApQk1VPiFYP6R_twkEjxnWJxtUPSDbXA8vo)

Let's get rid of the "x" button - the alert should not be dismissible.
Also move the alert under the GeneticTypeTabs because the user should be able to edit the germline if somatic is in review (and vice versa).
![image](https://github.com/user-attachments/assets/8e743999-a071-48ab-9bfc-4c7eeefa9b88)

gebbing12 commented:
> > I have added a banner to the top of the page. Please let me know if there are any improvements needed for the text or style. ![691733353841_ pic](https://private-user-images.githubusercontent.com/127727594/392596859-509ccc2a-e838-4da1-96e1-9428030977ea.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQwOTk4MTUsIm5iZiI6MTczNDA5OTUxNSwicGF0aCI6Ii8xMjc3Mjc1OTQvMzkyNTk2ODU5LTUwOWNjYzJhLWU4MzgtNGRhMS05NmUxLTk0MjgwMzA5NzdlYS5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIxM1QxNDE4MzVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05MjBhZDliZTYwZTY1ZTNkN2ZiZDI4NzVlZDUyZTg2YjY2YjQ2ZTJjYmU0MzFiNjE5MmQ3ZTg4ZTgzMjJiY2EzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.bwI98oCVApQk1VPiFYP6R_twkEjxnWJxtUPSDbXA8vo)
> 
> Let's get rid of the "x" button - the alert should not be dismissible. Also move the alert under the GeneticTypeTabs because the user should be able to edit the germline if somatic is in review (and vice versa). ![image](https://private-user-images.githubusercontent.com/59149377/395591368-8e743999-a071-48ab-9bfc-4c7eeefa9b88.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzNjg2MjMsIm5iZiI6MTczNDM2ODMyMywicGF0aCI6Ii81OTE0OTM3Ny8zOTU1OTEzNjgtOGU3NDM5OTktYTA3MS00OGFiLTliZmMtNGM3ZWVlZmE5Yjg4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE2VDE2NTg0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU5NTYyNGJkNmIyZDczMjdkNDYzZTNlMmViMTg4YzVmZWVlYzJiM2NmZDdlYzVkNDUwNGM5MDhlNDg1ODlhNmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.pYeMQQTtv8rpX6oWTxi2n4vBShtjGzV_85vEVGJ71IM)



> > I have added a banner to the top of the page. Please let me know if there are any improvements needed for the text or style. ![691733353841_ pic](https://private-user-images.githubusercontent.com/127727594/392596859-509ccc2a-e838-4da1-96e1-9428030977ea.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQwOTk4MTUsIm5iZiI6MTczNDA5OTUxNSwicGF0aCI6Ii8xMjc3Mjc1OTQvMzkyNTk2ODU5LTUwOWNjYzJhLWU4MzgtNGRhMS05NmUxLTk0MjgwMzA5NzdlYS5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIxM1QxNDE4MzVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05MjBhZDliZTYwZTY1ZTNkN2ZiZDI4NzVlZDUyZTg2YjY2YjQ2ZTJjYmU0MzFiNjE5MmQ3ZTg4ZTgzMjJiY2EzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.bwI98oCVApQk1VPiFYP6R_twkEjxnWJxtUPSDbXA8vo)
> 
> Let's get rid of the "x" button - the alert should not be dismissible. Also move the alert under the GeneticTypeTabs because the user should be able to edit the germline if somatic is in review (and vice versa). ![image](https://private-user-images.githubusercontent.com/59149377/395591368-8e743999-a071-48ab-9bfc-4c7eeefa9b88.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzNjg2MjMsIm5iZiI6MTczNDM2ODMyMywicGF0aCI6Ii81OTE0OTM3Ny8zOTU1OTEzNjgtOGU3NDM5OTktYTA3MS00OGFiLTliZmMtNGM3ZWVlZmE5Yjg4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE2VDE2NTg0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU5NTYyNGJkNmIyZDczMjdkNDYzZTNlMmViMTg4YzVmZWVlYzJiM2NmZDdlYzVkNDUwNGM5MDhlNDg1ODlhNmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.pYeMQQTtv8rpX6oWTxi2n4vBShtjGzV_85vEVGJ71IM)

Update of banner
![WeChatcbf7a0093376990bca3ecb126bdd317a](https://github.com/user-attachments/assets/b0bf7059-c163-477d-9a7d-540a07e2a95d)

calvinlu3 commented on src/main/webapp/app/pages/curation/header/ReadOnlyBanner.tsx:
```diff
@@ -26,15 +26,13 @@ const ReadOnlyBanner = ({ isGermline, hugoSymbol, firebaseDb }: IReadOnlyBanner)
       <BiInfoCircle size={25} className="me-2" />
-      <span>This page is currently in read-only mode because {meta.review.currentReviewer} is currently reviewing.</span>
-      <button type="button" className="btn-close" aria-label="Close" onClick={handleClose}></button>
+      <span>
+        {isGermline ? 'This Germline ' : 'This Somatic '}
+        is currently in read only mode because {meta.review.currentReviewer} is currently reviewing.
+      </span>
```
Maybe `The germline section is current in read only mode because {meta.review.currentReviewer} is currently reviewing.`

calvinlu3 commented on src/main/webapp/app/pages/curation/CurationPage.tsx:
```diff
@@ -86,6 +88,11 @@ export const CurationPage = (props: ICurationPageProps) => {
       const cleanupCallbacks: Unsubscribe[] = [];
+      cleanupCallbacks.push(
+        onValue(ref(props.firebaseDb, firebaseMetaCurrentReviewerPath), snapshot => {
+          props.setReadOnly(snapshot.val() ? true : false);
```
Can keep it concise using double negation
```suggestion
          props.setReadOnly(!!snapshot.val());
```

calvinlu3 commented on src/main/webapp/app/pages/curation/button/AddMutationButton.tsx:
```diff
@@ -7,9 +7,11 @@ const AddMutationButton: React.FunctionComponent<{
   showIcon?: boolean;
   showFullTitle?: boolean;
-}> = ({ showAddMutationModal, onClickHandler, showIcon = true, showFullTitle = false }) => {
+  disabled?: boolean;
```
I would keep it consistent with the `RCTButton` and just inject the CurationPage store into `AddMutationButton` and use the `readOnly` property for disabled. Same end result, but one less prop to pass in.

calvinlu3 commented on src/main/webapp/app/pages/curation/header/ReadOnlyBanner.tsx:
```diff
@@ -0,0 +1,47 @@
+const ReadOnlyBanner = ({ isGermline, hugoSymbol, firebaseDb }: IReadOnlyBanner) => {
+  const firebaseMetaPath = getFirebaseMetaGenePath(isGermline, hugoSymbol);
+  const [meta, setMeta] = useState<Meta>();
+  const [isVisible, setIsVisible] = useState(true);
```
Seems like this is not needed anymore because we always display the banner.

calvinlu3 commented on src/main/webapp/app/stores/curation-page.store.ts:
```diff
@@ -133,6 +137,10 @@ export class CurationPageStore {
   }
+
+  setReadOnly(isAnotherUserReviewing: boolean) {
+    this.readOnly = isAnotherUserReviewing ? true : false;
```
Already a boolean value, can just set as is.
```suggestion
    this.readOnly = isAnotherUserReviewing;
```

calvinlu3 reviewed (changes requested):
Looks good overall, just some minor comments. Thanks!

jfkonecn commented on src/main/webapp/app/pages/curation/header/ReadOnlyBanner.tsx:
```diff
@@ -0,0 +1,47 @@
+}
+
+const ReadOnlyBanner = ({ isGermline, hugoSymbol, firebaseDb }: IReadOnlyBanner) => {
+  const firebaseMetaPath = getFirebaseMetaGenePath(isGermline, hugoSymbol);
```
Please put this inside the `useEffect` and add `isGermline` and `hugoSymbol` to the dependency array.
If the path changes you'd would want the `useEffect` to be called again.

calvinlu3 commented on src/main/webapp/app/pages/curation/header/ReadOnlyBanner.tsx:
```diff
@@ -0,0 +1,48 @@
+
+  return meta ? (
+    <div className="alert alert-primary alert-dismissible fade show d-flex align-items-center" role="alert">
+      <BiInfoCircle size={25} className="me-2" />
```
```suggestion
      <BiInfoCircle size={DEFAULT_ICON_SIZE} className="me-2" />
```

calvinlu3 reviewed (approved):
LGTM! One last minor request.
